### ISSUE TITLE
[feature request] multiprocessing when collecting albums

### ISSUE THREAD

**Issue by thorsummoner** [saimn/sigal#353]
I notice the collecting albums phase is done with a single threaded for os.walk() read every file loop; and on systems with a single hard drive that's likely a reasonable thing to do, though on systems that can benefit from threaded reads, this is something of a painpoint/bottleneck (for me, waiting to have the gallery generated in minutes, not hours)

I'm willing to work on a codechange, would such a code change be welcome?

**Comment by saimn** [comment]
I never thought that the collecting albums step would be a bottleneck, but sigal was build for small/medium galleries, and as I'm quite happy with the current performance so I never really tried to optimize more. Do you have some numbers to get an idea of the performance in your case?

Also, as far as I remember, the collecting albums step is not reading the images. Almost all the processing is done in the multiprocessing loop (resizing images, creating thumbnails), and then the writing step can trigger some reads if some information is missing depending on the theme and settings.

So I certainly welcome efforts to improve the performance, though this must be balanced with the code complexity and speedup gain.

**Comment by thorsummoner** [comment]
i totally respect the scope of a small to medium gallery, i don't want to
push anything outside of the ambitions of this project. my frustrations led
me to try profiling the code and all that really suggested is that the
master thread which is the only one not sleeping was spending all it's time
reading from disk. I'll admit that system stats when i collected them show
100% active time on the virtual disk, I'm going to try to identify more
about that, it's a network volume so it's possible that moving the workload
to a system with bear metal oips might help me a bit

meanwhile, i can say with confidence that gnu `find` can list the directory
almost instantly, so there must be something more than just filesystem
enumeration at play. by the strace logs the main thread was opening every
file, reading 4096 bytes, and continuing to the next file. (as determined
by the heuristics of me making educated guesses)

this leads me to suspect that something about the image library under your
project is collecting data from the head of each file kind of thing; which
is why i hypothetisize making the gallery collection mutiprocessed work
could be useful for me.

i will play around some more, if i do propose any changes, I'll make an
effort to gate any specialized code i write to require intentional
activation via the argument parser.
I find myself preferring a different code style, which makes me hesitant to
get my hands dirty, i will also try to be a good contributor with adherence
to make-like like-like for any changes proposed. i would really prefer not
to make a change set drastic enough to require my own fork due to
internation concerns, but i blather on

Thanks for the feedback! and thanks for the welcoming words!

On Nov 16, 2018 3:50 PM, "Simon Conseil" <notifications@github.com> wrote:

I never thought that the collecting albums step would be a bottleneck, but
sigal was build for small/medium galleries, and as I'm quite happy with the
current performance so I never really tried to optimize more. Do you have
some numbers to get an idea of the performance in your case?

Also, as far as I remember, the collecting albums step is not reading the
images. Almost all the processing is done in the multiprocessing loop
(resizing images, creating thumbnails), and then the writing step can
trigger some reads if some information is missing depending on the theme
and settings.

So I certainly welcome efforts to improve the performance, though this must
be balanced with the code complexity and speedup gain.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/saimn/sigal/issues/353#issuecomment-439563059>, or mute
the thread

**Comment by thorsummoner** [comment]
just a note from the last successful run for me

Done.
Processed 1 images (434390 skipped) and 0 videos in 9990.87 seconds
(~3 hours)

time find Pictures/ >/dev/null
real    0m0.422s
user    0m0.188s
sys     0m0.232s

**Comment by saimn** [comment]
Good catch @Glandos , I forgot about this change. We could probably just add a setting to deactivate this.
